software/pio_hooks.py

```diff
@@ -342,18 +355,26 @@ def main():
         f.write('uint32_t build_timestamp(void);\n')
         f.write('const char *build_timestamp_hex_str(void);\n')
         f.write('const char *build_version_full_str(void);\n')
+        f.write('const char *build_info_str(void);\n')
 
     with open(os.path.join('src', 'build.cpp'), 'w', encoding='utf-8') as f:
         f.write('#include "build.h"\n')
         f.write('uint32_t build_timestamp(void) {{ return {}; }}\n'.format(timestamp))
         f.write('const char *build_timestamp_hex_str(void) {{ return "{:x}"; }}\n'.format(timestamp))
         f.write('const char *build_version_full_str(void) {{ return "{}.{}.{}-{:x}"; }}\n'.format(*version, timestamp))
+        f.write('const char *build_info_str(void) {{ return "git url: {}, git branch: {}, git commit id: {}"; }}\n'.format(git_url, branch_name, git_commit_id))
+        #f.write('const char *build_info_str(void) {{ return "git url: {}, git branch: {}, git commit id: {}, build flags: {}"; }}\n'.format(git_url, branch_name, git_commit_id, str(re.escape(build_flags))))
```

Comment:
I'd like to have the build flags in the build as well to make the build more reproducible, but could need some input on the if and how.

Reply:
You'd have to filter the build flags in some way if you want to do this. For example the last firmware I've built has the following flags:
```
-Os
-DBOARD_HAS_PSRAM
-DTF_NET_ENABLE=1
-DSNTP_GET_SERVERS_FROM_DHCP=1
-DOCPP_LOG_LEVEL=4
-DOCPP_PLATFORM_ESP32
-DOCPP_STATE_CALLBACKS
-DOCPP_SUPPORTED_FEATURE_PROFILES="\"Core,SmartCharging\""
-DOCPP_WEB_SOCKET_PING_INTERVAL=10
-DOCPP_WEB_SOCKET_PING_INTERVAL_REQUIRES_REBOOT=true
-DOCPP_METER_VALUES_ALIGNED_DATA_MAX_LENGTH=19
-DOCPP_DEFAULT_CLOCK_ALIGNED_DATA_INTERVAL=60
-DOCPP_DEFAULT_METER_VALUES_ALIGNED_DATA="\"Energy.Active.Export.Register,Energy.Active.Import.Register,Energy.Active.Export.Interval,Energy.Active.Import.Interval,Power.Active.Export,Power.Active.Import,Power.Offered,Power.Reactive.Export,Power.Reactive.Import,Power.Factor,Current.Import,Current.Export,Current.Offered,Voltage,Frequency\""
-DOCPP_METER_VALUES_SAMPLED_DATA_MAX_LENGTH=19
-DOCPP_DEFAULT_METER_VALUE_SAMPLE_INTERVAL=60
-DOCPP_DEFAULT_METER_VALUES_SAMPLED_DATA="\"Energy.Active.Export.Register,Energy.Active.Import.Register,Energy.Active.Export.Interval,Energy.Active.Import.Interval,Power.Active.Export,Power.Active.Import,Power.Offered,Power.Reactive.Export,Power.Reactive.Import,Power.Factor,Current.Import,Current.Export,Current.Offered,Voltage,Frequency\""
-DDEFAULT_WIFI_STA_ENABLE=true
-DDEFAULT_WIFI_STA_SSID="\"[REDACTED]\""
-DDEFAULT_WIFI_STA_PASSPHRASE="\"[REDACTED]\""
-DDEBUG_FS_ENABLE="\"true\""
```

For release builds, the build flags are checked in git anyway.

Reply:
> For release builds, the build flags are checked in git anyway.
Oh, how so? Where? That might be enough.

Reply:
They are part of the environments .ini file: https://github.com/Tinkerforge/esp32-firmware/blob/master/software/warp2.ini